### PR TITLE
Update build dependencies

### LINKED ISSUE #811: dollar and cent: variation in horizontal advance
FB reports that `dollar` and `cent` vary across the GRAD and/or ROND in horizontal advance

```
Ensure VFs with duplexed axes do not vary horizontal advance.
Check ID: com.google.fonts/check/varfont/duplexed_axis_reflow
🔥 GoogleSansFlex[GRAD,ROND,opsz,slnt,wdth,wght].ttf
🔥 FAIL
The following glyphs have variation in horizontal advance due to duplexed axes (GRAD or ROND): dollar and cent [code: duplexed-causes-reflow]
```

Tested data: https://github.com/googlefonts/googlesans-flex/actions/runs/7670878627

### PR DESCRIPTION
**Pull request by RickyDaMa**
Fixes #811, #904 

Will supercede the dependabot PRs

I'll wait and see if the pipeline passes, then if it does I'll look into TTX diffs. If there are non-negligible TTX diffs, I'll also produce a diffenator report

---

Pipeline regresses the following check: `com.google.fonts/check/dotted_circle`

The following glyphs could not be attached to the dotted circle glyph:
- uni0335
- uni0337

---

If possible, I'd also like to wait for https://github.com/googlefonts/gftools/commit/9d20701d8e57764c72c0f6d243ddd107499e50b6 to be released, though this is purely for the benefit of my personal/local workflow

### CODE REVIEW

**Comment by RickyDaMa** [comment]
Hey @MariannaPaszkowska, I've been looking into this regression with @madig's help, and this is what we've come up with:

It looks like previously this mark attachment was being done automatically by ufo2ft, however this behaviour was disabled [here](https://github.com/googlefonts/ufo2ft/pull/815) in order to more closely match Glyphs' behaviour (and hopefully Octavio's expectations also). Thus, what we'd recommend is for Octavio to review the anchorage with `uni25CC` (dotted circle) and ensure it passes this Fontbakery check. The alternative would be to explicitly opt in to ufo2ft's `DottedCircleFilter` which would do this automatically, but then we're not reviewing how good it looks (but it does pass the FB check)

Note: just because this is the only test that has regressed, this doesn't mean other anchor propogation won't be impacted by the change in ufo2ft. It would be good to review on the whole, not just related to dotted circle.

If you wanted to review, you can see a VF I produced with the new dependencies + `DottedCircleFilter`: 
[GoogleSansFlex.zip](https://github.com/googlefonts/googlesans-flex/files/14697354/GoogleSansFlex.GRAD.ROND.opsz.slnt.wdth.wght.zip). If you're interested in the VF from the updated dependencies (without the filter), you can download the one from the GitHub workflow run [here](https://github.com/googlefonts/googlesans-flex/actions/runs/8359712479?pr=902)

Once we've decided on our preferred approach here, I'll move forward with TTXing/diffenating

**Comment by MariannaPaszkowska** [comment]
Hmm.. this might point to a slightly bigger issue. 
From @octaviopardo 

_those two marks are actually bars we created to build components but in the glyphs file they are set as " do not export". They are the stroke short comb (used for glyphs like hbar or eth) and the slash short comb, for lslash. The only reason why we created them was to ensure that we compose those glyphs using components with automatic alignment. However stroke short comb was eventually decomposed in hbar and eth, before production, since the base glyphs (h and d) were also decomposed to slightly adjust their height and visually compensate the additional dark area on top._

It seems that we don't need those two glyphs in the fonts. Those should be decomposed and excluded.
Does it mean that something goes wrong during the conversion from .glyphspackage to UFOs? Does our pipeline ignore "do not export" settings? 
If so we need to think about how to address this.

Can we subset those in the statics?

**Comment by RickyDaMa** [comment]
Note to future self: this is probably caused by glyphsLib not writing `skipExportGlyphs` by default, which we could try enabling here: https://github.com/googlefonts/googlesans-flex/blob/76995cc95cfc3b83ab0abd7f6fe3dd73596e3ff4/scripts/glyphs_to_ds.py#L91

**Comment by RickyDaMa** [comment]
> Does our pipeline ignore "do not export" settings?

@MariannaPaszkowska yes, it did. Correcting this and doing a fresh import has added a bunch of `skipExportGlyphs` and solves this Fontbakery lint

Would you like me to include the changes to `skipExportGlyphs` as part of this PR, or should we just do a full re-import at some point?

<details>
<summary>Full list of new `skipExportGlyphs`</summary>

```
/_copyright-C
/_copyright-circle
/_dots-divide
/_registered-R
/_tail.Q
/eight.sinf
/five.sinf
/flig1
/flig2
/flig3
/four.sinf
/g.ss03
/gbreve.ss03
/gcommaaccent.ss03
/gdotaccent.ss03
/greaterequal-stroke
/ilig
/jlig
/nine.sinf
/notequal-slash
/one.sinf
/percentbar.tab
/seven.sinf
/six.sinf
/slashlongcomb.case
/slashshortcomb
/slashshortcomb.case
/strokelongcomb.case
/strokeshortcomb
/strokeshortcomb.case
/three.sinf
/tlig
/tlig2
/two.sinf
/zero.sinf
/zero.slash
```

</details>

---

Update from meeting 2024/03/26: don't update sources in this PR, just build dependencies. Let things get updated when an import is next done

**Comment by RickyDaMa** [comment]
Weird regression in latest gftools: build has gone from ~4 minutes to over 15, resulting in the timeout stopping the build

Update: can reproduce locally - seems to hang indefinitely instead of just a performance regression. Will look to make a bug report as I can't see anything on the gftools repo

**Comment by RickyDaMa** [comment]
Issue opened [here](https://github.com/googlefonts/gftools/issues/911), I've struggled to debug this with ninja orchestrating the Python programs being run

**Comment by simoncozens** [comment]
Taking a look.

**Comment by simoncozens** [comment]
On my computer, this runs

```
fontmake --output-path /var/folders/jp/4p0m9zvx2l739tdpflm38kv40000gn/T/tmp4oiko4g3 -o variable -m GoogleSansFlex.designspace --filter '...'  --filter FlattenComponentsFilter --filter DecomposeTransformedComponentsFilter --keep-overlaps
```

which takes a remarkable amount of time to even start to do anything, mainly because of the compatibility checker. Then it spits out a bunch of stuff, and then finally says:

```
INFO:ufo2ft.filters:Running interpolatable DecomposeTransformedComponentsIFilter
```

and there it stays, for longer than I cared to test. I have never come across this filter before; it seems to be a new thing as of two weeks ago.

**Comment by RickyDaMa** [comment]
Ah okay, so likely something else pulled in by the dependency update. I let it run for nearly an hour on my local machine and it made zero progress, so the size of our designspace must absolutely kill this filter. I'll look into it, cheers for the pointer Simon!

File: requirements.in [hunk 1]
```diff
@@ -6,4 +6,5 @@ matplotlib
 font-v
 
 # Latest stable version, for importing from .glyphs/.glyphspackage sources
-glyphslib>=6.4.1
+# glyphslib>=6.4.1
+glyphslib @ git+https://github.com/googlefonts/glyphsLib@no-width-linking-of-bracelayers
```
Comment (anthrotype) [inline requirements.in]: hm, not sure if it matters but, this branch is is [1 commit ahead of](https://github.com/googlefonts/glyphsLib/compare/main...no-width-linking-of-bracelayers), [29 commits behind](https://github.com/googlefonts/glyphsLib/compare/no-width-linking-of-bracelayers...main) glyphsLib `main`...

Comment (simoncozens) [reply requirements.in]: I was trying on head glyphslib/ufo2ft/fontmake and it was still taking forever.

**Comment by RickyDaMa** [comment]
@MariannaPaszkowska diffenator2 reports: [GS Flex dependency updates.zip](https://github.com/googlefonts/googlesans-flex/files/14917172/GS.Flex.dependency.updates.zip)

Shall review myself shortly also

Update: looks like the reports are completely broken, only showing notdefs. Need to investigate the cause of this further: both TTFs I used look fine

**Comment by RickyDaMa** [comment]
Hey @m4rc1e, I was wondering if you had a spare few minutes to take a look into this and help out

I've not been able to use the GitHub workflow you set up for us to run diffenator against the fonts this branch produces due to the issue I raised here: https://github.com/googlefonts/diffenator2/issues/134

I ran the report locally using the latest diffenator2 (0.3.8) by invoking:

```shell
diffenator2 diff --fonts-before "GoogleSansFlex[GRAD,ROND,opsz,slnt,wdth,wght] old deps.ttf" --fonts-after "fonts/variable/GoogleSansFlex[GRAD,ROND,opsz,slnt,wdth,wght].ttf" --out diffenator-report
```

(I ensured the output folder existed before running)

While the command ran to completion without issue (as far as I saw), the report is broken, not showing GS Flex on any of the generated pages. You can see the report attached to my previous comment

Any ideas what's going wrong here? I've poked at the TTFs I tested and they both seem absolutely fine, so I'm confused what's breaking diffenator

[fonts-before.zip](https://github.com/googlefonts/googlesans-flex/files/14931441/fonts-before.zip)
[fonts-after.zip](https://github.com/googlefonts/googlesans-flex/files/14931442/fonts-after.zip)

**Comment by anthrotype** [comment]
maybe it's the font file names containing spaces or special characters, try to rename them old.ttf and new.ttf

**Comment by RickyDaMa** [comment]
> maybe it's the font file names containing spaces or special characters, try to rename them old.ttf and new.ttf

...that works, thanks. I'm having a slow week, sorry!

**Comment by RickyDaMa** [comment]
@MariannaPaszkowska here's a working diffenator report: [GS Flex dependency updates.zip](https://github.com/googlefonts/googlesans-flex/files/14932038/GS.Flex.dependency.updates.zip)

Apologies for the delay!

**Comment by RickyDaMa** [comment]
Did some light reviewing myself (checking Regular, Black, and Thin), here's what I found (reviewed on Firefox on Windows)

On Regular, 9 subscript is less round on top and shifts down like a pixel:

9 subs shift

On Black, greater than or equal to gets a bit tighter:

geq shift

**Comment by m4rc1e** [comment]
Hey Ricky,

Apologies for the tool not working as expected. I'm able to replicate your results using a filename containing spaces. The issue is in the html template.


Screenshot 2024-04-10 at 16 35 48



`src: url(old-GoogleSansFlex[GRAD,ROND,opsz,slnt,wdth,wght] old deps.ttf);` needs quote marks around the filename.

`src: url("old-GoogleSansFlex[GRAD,ROND,opsz,slnt,wdth,wght] old deps.ttf");` does work. I'll update diff2 and fix the other issue you posted as well.

**Comment by belluzj** [comment]
@MariannaPaszkowska @chrissimpkins @schriftgestalt Following up on our meeting:

I can't update this PR to the latest fontmake (3.9.0) because of the glyphsLib pin to Georg's branch.

However, I'm not so sure that Georg's branch was necessary for the GS Flex release, because the pin of Georg's branch is only in this PR, but isn't in the main branch, or the tag that was used for v2.001 release: https://github.com/googlefonts/googlesans-flex/blob/v2.001/requirements.txt#L151

What problem was that branch fixing? Is the problem still relevant indeed? I found this in the meeting notes:

image

I'm sorry that we failed to review the branch and merge it in time for the v2.001 release, and find out only now. I'll double check whether what that branch does was really needed or not.

**Comment by RickyDaMa** [comment]
> However, I'm not so sure that Georg's branch was necessary for the GS Flex release, because the pin of Georg's branch is only in this PR, but isn't in the main branch, or the tag that was used for v2.001 release: https://github.com/googlefonts/googlesans-flex/blob/v2.001/requirements.txt#L151

Please see https://github.com/googlefonts/googlesans-flex/issues/811#issuecomment-2070129102

**Comment by belluzj** [comment]
@MariannaPaszkowska this is ready for review.

* The new dependencies fix an issue in v2.001, that is: glyphs marked "skip export" are now correctly excluded from the built TTF
* I've found the root cause of all the 1-pixel diffs in diffenator2: all glyph countours have been reversed with the new dependencies. But I think that the new direction is the correct one, so good news = this fixes an issue in the previous fonts, bad news = the diffenator diffs won't go away

Before (v2.001)
image


After (this PR):
image

**Comment by belluzj** [comment]
Looking into this outline direction problem:
* GS Flex v1 is OK (clockwise = TT direction)
* GS Flex v2 is flipped (counter-clockwise = PS direction)
  * Starting from the first Beta all the way to v2.001
* This branch is OK again

I tried to track down the reason why the outlines have been produced in the wrong direction for v2. On our side, the configuration variable to keep outline direction as per the sources has been here since the beginning: 

image

So it must be a regression in the dependencies used to build v2, prior to this update. I tried to debug the problem but I'm not sure it's worth figuring out which project had the regression, now that the latest versions seem to do the right thing again? Let me know if we should spend time to get to the bottom of this.

To prevent this from happening again in the future, we could work on an outline direction check in fontbakery's universal profile. There's an old issue from Nikolaus about this: https://github.com/fonttools/fontbakery/issues/2056

**Comment by MariannaPaszkowska** [comment]
@belluzj Thanks! Since it is already fixed and it's not easy to pin which of the tools has had that regression, let's drop the investigation and focus on finishing the dependency update. Let me know if I can help!

IMO, adding an outline direction check in Fontbakery's universal profile seems like a good idea. I'm surprised it wasn't included already. I vaguely recall such a test in the font validator tool from MS, and I thought all those were already included in FB... 🤔 

cc @simoncozens

**Comment by simoncozens** [comment]
Please tell me when gftools-builder does things you don't expect, I might be able to sort it out quickly. :-)

In this case, you are correctly passing `reverseOutlineDirection: false` but when I examine the `args = ` in the `build.ninja` file, I see the following:

```
    buildVariable GoogleSansFlex.designspace
  operation = buildVariable
  args = --filter ...  --filter FlattenComponentsFilter --filter $
      DecomposeTransformedComponentsFilter --keep-overlaps
  fontmake_type = -m
```

Funny, shouldn't args include ` --keep-direction`? 

```
        if self.config.get("reverseOutlineDirection"):
            args += " --keep-direction"
```

Well, that's completely wrong, we're only adding `--keep-direction` if `reverseOutlineDirection` is *true*!

I shall change it to

```
        if self.config.get("reverseOutlineDirection", True) is False:
            args += " --keep-direction"
```

and that will fix things for you...

**Comment by simoncozens** [comment]
This is fixed in the recently-released gftools 0.9.55.

**Comment by belluzj** [comment]
@simoncozens great, thanks for the fix! However @MariannaPaszkowska we have a problem, it means the outlines will go in the wrong direction again after Simon's fix. So, I'm starting to think the outlines may be drawn the PS way in the Glyphs sources. I'll go and check that, and if that's the case, we may want to indeed reverse them in our pipeline.

Edit: the sources are in PS direction indeed, I suppose since version 2. I started a discussion in the chat about next step (reverse outlines in sources or reverse them in pipeline basically).

**Comment by schriftgestalt** [comment]
Yes. Outlines in Glyphs are always in PS direction. There are numerous reasons to do so (I explain if needed). So when you export to TrueType, the outlines always need to be reversed.

**Comment by simoncozens** [comment]
I'm working on a fontbakery check. The weird thing is that, in my test, fontmake seems to produce the same `glyf`table whether the path is clockwise or anticlockwise in the sources.

**Comment by belluzj** [comment]
Thanks @schriftgestalt, according to this I'll change the pipeline to reverse the outline direction. 

I found in the meeting agenda that this issue was raised in August 2023 and it appears we forgot to investigate it until now, I apologize for the oversight.

image

**Comment by MariannaPaszkowska** [comment]
Thank you, everyone, path direction mystery solved! 
@simoncozens many thanks for your help! This is weird, indeed. Please keep us updated!

**Comment by anthrotype** [comment]
> The weird thing is that, in my test, fontmake seems to produce the same glyftable whether the path is clockwise or anticlockwise in the sources.

how's that possible? Have you found out?

**Comment by anthrotype** [comment]
By the way, don't know if you know, if your sources only contain TrueType quadratic curves and no cubic beziers, you can tell fontmake to skip cu2qu step altogether. There's a `--ttf-curves` option that controls this:

https://github.com/googlefonts/fontmake/blob/8305773a3a495750832b2f1929eeb83c9cefdae5/Lib/fontmake/__main__.py#L449-L459

basically fontmake by default runs cu2qu to convert all cubics to quads; if there are only quads like in your case it's no-op but still may take some time to process all glyphs. If you set it to `--ttf-curves=keep-quad` it will attempt to compile the curves as is (and potentially fail with an error if they actually aren't as told).

**Comment by simoncozens** [comment]
Yeah, seems like the remove overlaps filter fixes the path direction.

**Comment by anthrotype** [comment]
hm so you have TrueType paths in your sources _with_ overlaps to be removed at build time? Yeah, skia-pathops (the only backend that supports quadratic curves, booleanOperations doesn't) has a fix_winding option to compute non-zero winding of contours because the underlying PathOps will return them as even-odd...
I'll look into it

**Comment by anthrotype** [comment]
ufo2ft removeOverlapsFilter calls pathops.operations.union function but doesn't set the `clockwise` parameter so it defaults to False (which means it will always set it to counterclockwise).
https://github.com/googlefonts/ufo2ft/blob/8e9e6eb66a2d22714c9da9544bdbfe1aa44d0cf0/Lib/ufo2ft/filters/removeOverlaps.py#L47
https://github.com/fonttools/skia-pathops/blob/815070e964334ec6cdf453f2d859fe2db49d9b06/src/python/pathops/operations.py#L27
It should instead set that parameter based on the value of `reverseDirection` and whether it's building a TTF or an OTF...

**Comment by simoncozens** [comment]
> hm so you have TrueType paths in your sources with overlaps to be removed at build time? 

I'm not directly working on GSF, so my comment wasn't about that - I was trying to come up with a test for the fontbakery direction check that Marianna asked for, and was surprised that I couldn't produce one because fontmake was correcting my path directions.

**Comment by MariannaPaszkowska** [comment]
The updated diffenator2 doesn't report anymore differences. Thanks, @m4rc1e Reports: https://github.com/googlefonts/googlesans-flex/actions/runs/9001941157 !

**Pull request merged**